Test Full Pipeline For BMI detection using extra features extracted as well

In [30]:
# conda create -n bmi-predictor python=3.10 -y
# conda activate bmi-predictor

# # For MPS on macOS with ARM (M1/M2/M3)
# pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu

# pip install facenet-pytorch
# pip install pandas scikit-learn matplotlib tqdm jupyter notebook
# pip install opencv-python pillow

### Data Preprocessing

In [31]:
# ===============================================
# STEP 0: Imports and Setup
# ===============================================
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import optim
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms
from facenet_pytorch import InceptionResnetV1, MTCNN
import pandas as pd
import numpy as np
from PIL import Image
from pathlib import Path
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from scipy.stats import pearsonr
from tqdm import tqdm 

# DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f'Core count: {os.cpu_count()}')
torch.set_num_threads(os.cpu_count())  # Use all available CPU cores
DEVICE = torch.device("cpu")  # Force CPU usage for portability
print("Using device:", DEVICE)

Core count: 10
Using device: cpu


In [32]:
# Paths
IMAGE_DIR = "data/Images"
existing_image_files = os.listdir(IMAGE_DIR)

# Load CSV
data_df = pd.read_csv("data/data.csv").drop(columns=["Unnamed: 0"])
print(f"Loaded {data_df.shape[0]} rows from CSV")

# Filter out rows with missing image files
data_df = data_df[data_df["name"].isin(existing_image_files)]
print(f"Filtered to {data_df.shape[0]} rows with existing images")

del existing_image_files
data_df.head(5)

Loaded 4206 rows from CSV
Filtered to 3962 rows with existing images


bmi  gender  is_training       name
0  34.207396    Male            1  img_0.bmp
1  26.453720    Male            1  img_1.bmp
2  34.967561  Female            1  img_2.bmp
3  22.044766  Female            1  img_3.bmp
6  25.845588  Female            1  img_6.bmp

In [33]:
features_df = pd.read_csv('data/Feature_Add.csv')
features_df.head()

jaw_width  face_height  cheekbone_width  nose_width  mouth_width  \
0  237.002110   210.287898       237.008439   52.009614   120.016666   
1  226.035395   200.010000       220.002273   51.000000   104.076895   
2  241.101638   209.193690       239.075302   42.011903    87.005747   
3  229.490741   198.494332       229.176788   53.037722    80.056230   
4  235.552542   201.613492       235.766834   48.041649    77.524190   

   mouth_height  eye_distance  left_eye_width  right_eye_width  \
0     58.034473     62.008064       46.173586        42.107007   
1     38.052595     62.072538       43.011626        41.048752   
2     32.062439     65.000000       42.107007        43.011626   
3     12.000000     54.083269       42.011903        41.012193   
4     27.892651     60.299254       34.014703        41.194660   

   face_width_to_height  mouth_to_nose_ratio               image        race  \
0              1.127036             2.307586  img_1066_face0.jpg       White   
1              1.130120             2.040723  img_1585_face0.jpg       White   
2              1.152528             2.070978  img_3852_face0.jpg       White   
3              1.156158             1.509421  img_1857_face0.jpg  East Asian   
4              1.168337             1.613687  img_4196_face0.jpg       White   

  pred_gender    age       key        bmi  gender  is_training          name  
0      Female  20-29  img_1066  41.191406  Female            1  img_1066.bmp  
1        Male  20-29  img_1585  24.658895    Male            1  img_1585.bmp  
2      Female  20-29  img_3852  39.151259  Female            0  img_3852.bmp  
3        Male  30-39  img_1857  25.845588    Male            1  img_1857.bmp  
4        Male  20-29  img_4196  36.243556    Male            0  img_4196.bmp

In [34]:
df = pd.merge(data_df, features_df, on=['gender', 'bmi', 'name', 'is_training'], how='inner')
df.drop(columns=['key', 'image', 'gender'], inplace=True)
print("Combined dataset shape:", df.shape)
df.head()

Combined dataset shape: (3712, 17)


bmi  is_training       name   jaw_width  face_height  \
0  34.207396            1  img_0.bmp  242.008264   176.045449   
1  26.453720            1  img_1.bmp  227.140925   185.067555   
2  34.967561            1  img_2.bmp  269.185809   217.057596   
3  22.044766            1  img_3.bmp  244.018442   216.148097   
4  25.845588            1  img_6.bmp  238.838858   196.063765   

   cheekbone_width  nose_width  mouth_width  mouth_height  eye_distance  \
0       239.002092   50.000000    97.020616     18.027756     62.008064   
1       220.056811   48.041649    88.051122     22.022716     59.008474   
2       266.030073   48.000000   103.019416     51.000000     59.033889   
3       239.002092   44.045431   102.044108     50.039984     59.008474   
4       236.541751   48.041649    93.085982     24.020824     69.260378   

   left_eye_width  right_eye_width  face_width_to_height  mouth_to_nose_ratio  \
0       44.045431        43.046487              1.374692             1.940412   
1       41.012193        43.185646              1.227341             1.832808   
2       42.011903        44.045431              1.240158             2.146238   
3       41.000000        41.012193              1.128941             2.316792   
4       48.259714        47.010637              1.218169             1.937610   

    race pred_gender    age  
0  White        Male  30-39  
1  White        Male  20-29  
2  White      Female  20-29  
3  White      Female  20-29  
4  White      Female  10-19

In [35]:
# Normalize target variable
bmi_mean = df['bmi'].mean()
bmi_std = df['bmi'].std()
df['bmi_norm'] = (df['bmi'] - bmi_mean) / bmi_std
df.head()

bmi  is_training       name   jaw_width  face_height  \
0  34.207396            1  img_0.bmp  242.008264   176.045449   
1  26.453720            1  img_1.bmp  227.140925   185.067555   
2  34.967561            1  img_2.bmp  269.185809   217.057596   
3  22.044766            1  img_3.bmp  244.018442   216.148097   
4  25.845588            1  img_6.bmp  238.838858   196.063765   

   cheekbone_width  nose_width  mouth_width  mouth_height  eye_distance  \
0       239.002092   50.000000    97.020616     18.027756     62.008064   
1       220.056811   48.041649    88.051122     22.022716     59.008474   
2       266.030073   48.000000   103.019416     51.000000     59.033889   
3       239.002092   44.045431   102.044108     50.039984     59.008474   
4       236.541751   48.041649    93.085982     24.020824     69.260378   

   left_eye_width  right_eye_width  face_width_to_height  mouth_to_nose_ratio  \
0       44.045431        43.046487              1.374692             1.940412   
1       41.012193        43.185646              1.227341             1.832808   
2       42.011903        44.045431              1.240158             2.146238   
3       41.000000        41.012193              1.128941             2.316792   
4       48.259714        47.010637              1.218169             1.937610   

    race pred_gender    age  bmi_norm  
0  White        Male  30-39  0.191720  
1  White        Male  20-29 -0.762228  
2  White      Female  20-29  0.285244  
3  White      Female  20-29 -1.304670  
4  White      Female  10-19 -0.837048

In [36]:
#### General information ####
print(f"Entire data shape: {df.shape}")
gender_total = df['pred_gender'].value_counts() / len(df)
gender_total = gender_total.reset_index()
gender_total.rename(columns={'count': 'percentage'}, inplace=True)
print(f"Gender Distribution Over Entire Dataset:\n{gender_total}\n")

# Split into training and testing sets
training_data_df = df[df['is_training'] == 1]
testing_data_df = df[df['is_training'] == 0]

gender_training_data_df = training_data_df['pred_gender'].value_counts() / len(training_data_df)
gender_training_data_df = gender_training_data_df.reset_index()
gender_training_data_df.rename(columns={'count': 'percentage'}, inplace=True)
print(f"Training data shape: {training_data_df.shape}")
print(f"Gender Distribution Over Training Dataset:\n{gender_training_data_df}\n")

gender_testing_data_df = testing_data_df['pred_gender'].value_counts() / len(testing_data_df)
gender_testing_data_df = gender_testing_data_df.reset_index()
gender_testing_data_df.rename(columns={'count': 'percentage'}, inplace=True)
print(f"Testing data shape: {testing_data_df.shape}")
print(f"Gender Distribution Over Testing Dataset:\n{gender_testing_data_df}\n")

del data_df, features_df, gender_total, training_data_df, testing_data_df, gender_training_data_df, gender_testing_data_df

Entire data shape: (3712, 18)
Gender Distribution Over Entire Dataset:
  pred_gender  percentage
0        Male    0.581088
1      Female    0.418912

Training data shape: (3008, 18)
Gender Distribution Over Training Dataset:
  pred_gender  percentage
0        Male    0.590093
1      Female    0.409907

Testing data shape: (704, 18)
Gender Distribution Over Testing Dataset:
  pred_gender  percentage
0        Male    0.542614
1      Female    0.457386



In [37]:
### One Hot Encode ###
race_dummies = pd.get_dummies(df['race'], prefix='race').astype(int)
gender_dummies = pd.get_dummies(df['pred_gender'], prefix='gender').astype(int)
age_dummies = pd.get_dummies(df['age'], prefix='age').astype(int)

# Fix column names immediately after get_dummies
race_dummies.columns = [col.replace(" ", "_") for col in race_dummies.columns]
gender_dummies.columns = [col.replace(" ", "_") for col in gender_dummies.columns]
age_dummies.columns = [col.replace(" ", "_") for col in age_dummies.columns]

# Combine with original numeric features
df_final = pd.concat([
    df.drop(['race', 'pred_gender', 'age'], axis=1),
    race_dummies,
    gender_dummies,
    age_dummies
], axis=1)


del df, race_dummies, gender_dummies, age_dummies
df_final.head()

bmi  is_training       name   jaw_width  face_height  \
0  34.207396            1  img_0.bmp  242.008264   176.045449   
1  26.453720            1  img_1.bmp  227.140925   185.067555   
2  34.967561            1  img_2.bmp  269.185809   217.057596   
3  22.044766            1  img_3.bmp  244.018442   216.148097   
4  25.845588            1  img_6.bmp  238.838858   196.063765   

   cheekbone_width  nose_width  mouth_width  mouth_height  eye_distance  ...  \
0       239.002092   50.000000    97.020616     18.027756     62.008064  ...   
1       220.056811   48.041649    88.051122     22.022716     59.008474  ...   
2       266.030073   48.000000   103.019416     51.000000     59.033889  ...   
3       239.002092   44.045431   102.044108     50.039984     59.008474  ...   
4       236.541751   48.041649    93.085982     24.020824     69.260378  ...   

   race_Latino_Hispanic  race_White  gender_Female  gender_Male  age_10-19  \
0                     0           1              0            1          0   
1                     0           1              0            1          0   
2                     0           1              1            0          0   
3                     0           1              1            0          0   
4                     0           1              1            0          1   

   age_20-29  age_3-9  age_30-39  age_40-49  age_50-59  
0          0        0          1          0          0  
1          1        0          0          0          0  
2          1        0          0          0          0  
3          1        0          0          0          0  
4          0        0          0          0          0  

[5 rows x 27 columns]

In [38]:
df_final.columns

Index(['bmi', 'is_training', 'name', 'jaw_width', 'face_height',
       'cheekbone_width', 'nose_width', 'mouth_width', 'mouth_height',
       'eye_distance', 'left_eye_width', 'right_eye_width',
       'face_width_to_height', 'mouth_to_nose_ratio', 'bmi_norm', 'race_Black',
       'race_East_Asian', 'race_Latino_Hispanic', 'race_White',
       'gender_Female', 'gender_Male', 'age_10-19', 'age_20-29', 'age_3-9',
       'age_30-39', 'age_40-49', 'age_50-59'],
      dtype='object')

In [39]:
extra_feature_cols = [col for col in df_final.columns if col not in ['name', 'bmi', 'bmi_norm' 'is_training']]
print(df_final[extra_feature_cols].dtypes[df_final[extra_feature_cols].dtypes == "object"])

Series([], dtype: object)


### Face Detection & Alignment using MTCNN

In [40]:
# ===============================================
# STEP X: Precompute and Cache Face Embeddings
# ===============================================
from facenet_pytorch import InceptionResnetV1, MTCNN
from torchvision import transforms
import torch
from PIL import Image
import os
from tqdm import tqdm
import numpy as np

IMAGE_DIR = "data/Images"
EMBEDDING_CACHE_DIR = "data/embeddings"
os.makedirs(EMBEDDING_CACHE_DIR, exist_ok=True)

device = torch.device("cpu")
# mtcnn = MTCNN(image_size=160, margin=20, post_process=True, device=device)
mtcnn = MTCNN(image_size=224, margin=20, post_process=True, device=device)
embedder = InceptionResnetV1(pretrained="vggface2").eval().to(device)

def extract_face_embedding(image_path):
    img = Image.open(image_path).convert("RGB")
    face = mtcnn(img)
    if face is None:
        # return None
        return torch.zeros(512).numpy()
    face = face.unsqueeze(0).to(device)
    with torch.no_grad():
        embedding = embedder(face).squeeze(0).cpu().numpy()
    return embedding

generate_emb = input("Generate embeddings?: Y or N").lower().strip() or "n"

if generate_emb == 'y':
    # Generate embeddings
    for idx, row in tqdm(df_final.iterrows(), total=len(df_final), desc="Caching embeddings"):
        image_name = row["name"]
        image_path = os.path.join(IMAGE_DIR, image_name)
        save_path = os.path.join(EMBEDDING_CACHE_DIR, image_name.replace(".bmp", ".npy"))

        # if not os.path.exists(save_path):
        try:
            embedding = extract_face_embedding(image_path)
            if embedding is not None:
                np.save(save_path, embedding)
        except Exception as e:
            print(f"Failed {image_name}: {e}")
else: 
    print("Will use saved embeddings from prior process")

Will use saved embeddings from prior process


In [41]:
# Remove images we were not able to create embeddings 
from pathlib import Path
df_final = df_final[df_final["name"].apply(lambda x: Path("data/embeddings") / x.replace(".bmp", ".npy")).apply(Path.exists)].reset_index(drop=True)
print(df_final.shape)

(3712, 27)


In [42]:
class BMIDataset(Dataset):
    def __init__(self, dataframe, image_dir, embedding_dir="data/embeddings"):
        self.dataframe = dataframe.reset_index(drop=True)
        self.image_dir = image_dir
        self.embedding_dir = embedding_dir

        # Identify tabular feature columns
        self.extra_feature_cols = [
            col for col in dataframe.columns
            if col not in ['name', 'bmi', 'bmi_norm', 'is_training']
        ]

    def __len__(self):
        return len(self.dataframe)

    def __getitem__(self, idx):
        row = self.dataframe.iloc[idx]
        image_name = row["name"]
        # bmi = torch.tensor(row["bmi"], dtype=torch.float32) #non normalized value 
        bmi = torch.tensor(row['bmi_norm'], dtype=torch.float32) # use normalized value
        # Load cached face embedding
        embedding_path = os.path.join(self.embedding_dir, image_name.replace(".bmp", ".npy"))
        embedding = np.load(embedding_path)
        face_tensor = torch.tensor(embedding, dtype=torch.float32)

        # Load extra tabular features
        extra_features = torch.tensor(row[self.extra_feature_cols].values.astype(np.float32))

        return face_tensor, extra_features, bmi

In [64]:
# ===============================================
# STEP 5: Model Definition (Widened Multitask)
# ===============================================
class BMIMultitaskModel(nn.Module):
    def __init__(self, extra_feature_dim):
        super().__init__()
        self.backbone = None  # no longer needed with cached embeddings

        # Wider shared layer
        self.shared = nn.Sequential(
            nn.Linear(512 + extra_feature_dim, 384),
            nn.ReLU(),
            nn.BatchNorm1d(384),
            nn.Dropout(0.3),
            nn.Linear(384, 256),
            nn.ReLU(),
            nn.BatchNorm1d(256),
            nn.Dropout(0.3),
        )

        # BMI prediction head
        self.bmi_head = nn.Sequential(
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 1)
        )

        # Age group regression head
        self.age_head = nn.Sequential(
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 1)
        )

        # Gender classification head (binary)
        self.gender_head = nn.Sequential(
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 1)  # Sigmoid for binary classification
        )

    def forward(self, face_embed, extra_features):
        x = torch.cat([face_embed, extra_features], dim=1)
        x = self.shared(x)

        bmi = self.bmi_head(x).squeeze(1)
        age = self.age_head(x).squeeze(1)
        gender = self.gender_head(x).squeeze(1)  # sigmoid if binary

        return bmi, age, gender

In [65]:
# ===============================================
# STEP 6: Prepare Data Loaders
# ===============================================

# Obtain values to normalize target variable

X_train = df_final[df_final['is_training'] == 1]
test_df = df_final[df_final['is_training'] == 0]

# Create validation set
validation_set_percent = 0.3

train_df, validation_df = train_test_split(X_train, test_size=0.3, random_state=42)

print(f"Training set shape: {train_df.shape}")
print(f"Validation set shape: {validation_df.shape}")
print(f"Number of Males in training set shape: {train_df['gender_Male'].sum()}; {train_df['gender_Male'].sum()/len(train_df) * 100:.2f}%")
print(f"Number of Females in training set shape: {train_df['gender_Female'].sum()}; {train_df['gender_Female'].sum()/len(train_df) * 100:.2f}%")
print(f"Number of Males in validation set shape: {validation_df['gender_Male'].sum()}; {validation_df['gender_Male'].sum()/len(validation_df) * 100:.2f}%")
print(f"Number of Females in validation set shape: {validation_df['gender_Female'].sum()}; {validation_df['gender_Female'].sum()/len(validation_df) * 100:.2f}%")

train_dataset = BMIDataset(train_df, image_dir=IMAGE_DIR)
val_dataset = BMIDataset(validation_df, image_dir=IMAGE_DIR)
test_dataset = BMIDataset(test_df, image_dir=IMAGE_DIR)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True, num_workers=0)
validation_loader = DataLoader(val_dataset, batch_size=32, shuffle=True, num_workers=0)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False, num_workers=0)

Training set shape: (2105, 27)
Validation set shape: (903, 27)
Number of Males in training set shape: 1241; 58.95%
Number of Females in training set shape: 864; 41.05%
Number of Males in validation set shape: 534; 59.14%
Number of Females in validation set shape: 369; 40.86%


In [66]:
# Configure paths for best model
MODEL_PATH = "saved_models/best_bmi_prediction_model.pth"
os.makedirs(os.path.dirname(MODEL_PATH), exist_ok=True)

In [67]:
# ===============================================
# STEP 7: Training Loop
# ===============================================

def evaluate(model, criterion, dataloader):
    model.eval()
    val_loss = 0.0
    all_preds, all_labels = [], []
    with torch.no_grad():
        for face, extra, bmi in tqdm(dataloader, desc="Evaluating", leave=False):
            face = face.to(DEVICE, non_blocking=True)
            extra = extra.to(DEVICE, non_blocking=True)
            bmi = bmi.to(DEVICE, non_blocking=True)

            pred_bmi, _, _ = model(face, extra)
            loss = criterion(pred_bmi, bmi)
            val_loss += loss.item()
            all_preds.append(pred_bmi.cpu().numpy())
            all_labels.append(bmi.cpu().numpy())
    
    avg_val_loss = val_loss / len(dataloader)
    
    preds = np.concatenate(all_preds)
    labels = np.concatenate(all_labels)
    
    # Reverse transformation from normalized bmi values
    preds = preds * bmi_std + bmi_mean
    labels = labels * bmi_std + bmi_mean
    
    # Evaluate performance
    mae = mean_absolute_error(labels, preds)
    r2 = r2_score(labels, preds)
    pearson_corr, _ = pearsonr(labels, preds)

    print(f"Evaluation: MAE={mae:.2f}, R²={r2:.3f}, Pearson r={pearson_corr:.3f}")
    return avg_val_loss, mae, r2, pearson_corr

def train_epoch(model, optimizer, criterion, dataloader):
    model.train()
    total_loss = 0
    for face, extra, bmi in tqdm(dataloader, desc="Training", leave=False):
        # face, extra, bmi = face.to(DEVICE), extra.to(DEVICE), bmi.to(DEVICE)
        face = face.to(DEVICE, non_blocking=True)
        extra = extra.to(DEVICE, non_blocking=True)
        bmi = bmi.to(DEVICE, non_blocking=True)
        
        optimizer.zero_grad()
        pred_bmi, _, _ = model(face, extra)  # Only use BMI for loss
        loss = criterion(pred_bmi, bmi)
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    
    return total_loss / len(dataloader)

def train_model(model, optimizer, criterion, epochs=150):
    best_val_loss = float('inf')
    
    for epoch in range(epochs):
        loss = train_epoch(model, optimizer, criterion, train_loader)
        avg_val_loss, val_mae, r2_value, pearson_corr = evaluate(model, criterion, validation_loader)
        print(f"Epoch [{epoch+1}/{epochs}]: Train Loss={loss:.4f}, Val Loss={avg_val_loss:.4f}, Val MAE={val_mae:.3f}, Val R^2={r2_value:.3f}, Val Pearson Coefficient={pearson_corr:.3f}")
        
        # Save best model
        if avg_val_loss < best_val_loss:
            print(f"Prior best: {best_val_loss}")
            best_val_loss = avg_val_loss
            print(f"Current best: {best_val_loss}")
            torch.save(model.state_dict(), MODEL_PATH)
            print(f"Best model saved to: {MODEL_PATH}")

def load_best_model(extra_feature_dim):
    model = BMIMultitaskModel(extra_feature_dim=extra_feature_dim).to(DEVICE)
    checkpoint = torch.load(MODEL_PATH, map_location=DEVICE)
    model.load_state_dict(checkpoint)
    return model

In [73]:
# Train model
is_train_model = input("Train new model? Y or N\n Default: Use best saved model").lower().strip() or "n"

model = BMIMultitaskModel(extra_feature_dim=len(train_dataset.extra_feature_cols)).to(DEVICE)
# optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4, weight_decay=1e-5)
criterion = nn.MSELoss()

if is_train_model == 'y':
    print("Beginning training")
    train_model(model, optimizer, criterion)
    print("Training Complete")
else:
    print("Loading saved best model")
# Get extra features dimension
extra_feature_dim = len(test_dataset.extra_feature_cols)
best_model = load_best_model(extra_feature_dim)

Beginning training


Evaluation: MAE=5.91, R²=0.159, Pearson r=0.409
Epoch [1/150]: Train Loss=0.8675, Val Loss=0.8399, Val MAE=5.913, Val R^2=0.159, Val Pearson Coefficient=0.409
Prior best: inf
Current best: 0.8399127002420097
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.78, R²=0.195, Pearson r=0.445
Epoch [2/150]: Train Loss=0.8155, Val Loss=0.8190, Val MAE=5.781, Val R^2=0.195, Val Pearson Coefficient=0.445
Prior best: 0.8399127002420097
Current best: 0.8190476688845404
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.59, R²=0.180, Pearson r=0.466
Epoch [3/150]: Train Loss=0.7717, Val Loss=0.8265, Val MAE=5.591, Val R^2=0.180, Val Pearson Coefficient=0.466


Evaluation: MAE=5.53, R²=0.232, Pearson r=0.483
Epoch [4/150]: Train Loss=0.7547, Val Loss=0.7704, Val MAE=5.532, Val R^2=0.232, Val Pearson Coefficient=0.483
Prior best: 0.8190476688845404
Current best: 0.7704322009251036
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.46, R²=0.237, Pearson r=0.499
Epoch [5/150]: Train Loss=0.7388, Val Loss=0.7693, Val MAE=5.462, Val R^2=0.237, Val Pearson Coefficient=0.499
Prior best: 0.7704322009251036
Current best: 0.7692733211763974
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.35, R²=0.231, Pearson r=0.524
Epoch [6/150]: Train Loss=0.6993, Val Loss=0.7908, Val MAE=5.350, Val R^2=0.231, Val Pearson Coefficient=0.524


Evaluation: MAE=5.37, R²=0.290, Pearson r=0.555
Epoch [7/150]: Train Loss=0.6805, Val Loss=0.7251, Val MAE=5.372, Val R^2=0.290, Val Pearson Coefficient=0.555
Prior best: 0.7692733211763974
Current best: 0.7251461785415123
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.05, R²=0.331, Pearson r=0.588
Epoch [8/150]: Train Loss=0.6625, Val Loss=0.6792, Val MAE=5.050, Val R^2=0.331, Val Pearson Coefficient=0.588
Prior best: 0.7251461785415123
Current best: 0.6791917517267424
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=4.87, R²=0.371, Pearson r=0.615
Epoch [9/150]: Train Loss=0.6396, Val Loss=0.6254, Val MAE=4.870, Val R^2=0.371, Val Pearson Coefficient=0.615
Prior best: 0.6791917517267424
Current best: 0.6253500950747523
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=6.63, R²=-0.136, Pearson r=0.613
Epoch [10/150]: Train Loss=0.6271, Val Loss=1.1421, Val MAE=6.631, Val R^2=-0.136, Val Pearson Coefficient=0.613


Evaluation: MAE=4.78, R²=0.364, Pearson r=0.626
Epoch [11/150]: Train Loss=0.6182, Val Loss=0.6394, Val MAE=4.779, Val R^2=0.364, Val Pearson Coefficient=0.626


Evaluation: MAE=4.85, R²=0.385, Pearson r=0.625
Epoch [12/150]: Train Loss=0.6140, Val Loss=0.6196, Val MAE=4.854, Val R^2=0.385, Val Pearson Coefficient=0.625
Prior best: 0.6253500950747523
Current best: 0.6195562923776692
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=4.69, R²=0.401, Pearson r=0.636
Epoch [13/150]: Train Loss=0.5771, Val Loss=0.5966, Val MAE=4.693, Val R^2=0.401, Val Pearson Coefficient=0.636
Prior best: 0.6195562923776692
Current best: 0.5965532300801113
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.25, R²=0.350, Pearson r=0.642
Epoch [14/150]: Train Loss=0.6021, Val Loss=0.6691, Val MAE=5.248, Val R^2=0.350, Val Pearson Coefficient=0.642


Evaluation: MAE=4.72, R²=0.418, Pearson r=0.649
Epoch [15/150]: Train Loss=0.5989, Val Loss=0.5943, Val MAE=4.724, Val R^2=0.418, Val Pearson Coefficient=0.649
Prior best: 0.5965532300801113
Current best: 0.5943373246439572
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=4.97, R²=0.392, Pearson r=0.645
Epoch [16/150]: Train Loss=0.5838, Val Loss=0.6086, Val MAE=4.973, Val R^2=0.392, Val Pearson Coefficient=0.645


Evaluation: MAE=4.69, R²=0.382, Pearson r=0.647
Epoch [17/150]: Train Loss=0.5774, Val Loss=0.6268, Val MAE=4.685, Val R^2=0.382, Val Pearson Coefficient=0.647


Evaluation: MAE=4.99, R²=0.294, Pearson r=0.640
Epoch [18/150]: Train Loss=0.5823, Val Loss=0.7198, Val MAE=4.991, Val R^2=0.294, Val Pearson Coefficient=0.640


Evaluation: MAE=5.62, R²=0.120, Pearson r=0.642
Epoch [19/150]: Train Loss=0.5868, Val Loss=0.8884, Val MAE=5.618, Val R^2=0.120, Val Pearson Coefficient=0.642


Evaluation: MAE=4.66, R²=0.411, Pearson r=0.647
Epoch [20/150]: Train Loss=0.5790, Val Loss=0.5866, Val MAE=4.663, Val R^2=0.411, Val Pearson Coefficient=0.647
Prior best: 0.5943373246439572
Current best: 0.5866291656576353
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=7.67, R²=-0.181, Pearson r=0.639
Epoch [21/150]: Train Loss=0.5482, Val Loss=1.2063, Val MAE=7.665, Val R^2=-0.181, Val Pearson Coefficient=0.639


Evaluation: MAE=4.98, R²=0.284, Pearson r=0.652
Epoch [22/150]: Train Loss=0.5684, Val Loss=0.7317, Val MAE=4.984, Val R^2=0.284, Val Pearson Coefficient=0.652


Evaluation: MAE=6.03, R²=0.194, Pearson r=0.648
Epoch [23/150]: Train Loss=0.5731, Val Loss=0.8106, Val MAE=6.034, Val R^2=0.194, Val Pearson Coefficient=0.648


Evaluation: MAE=4.78, R²=0.406, Pearson r=0.640
Epoch [24/150]: Train Loss=0.5680, Val Loss=0.5962, Val MAE=4.777, Val R^2=0.406, Val Pearson Coefficient=0.640


Evaluation: MAE=4.83, R²=0.416, Pearson r=0.654
Epoch [25/150]: Train Loss=0.5720, Val Loss=0.5807, Val MAE=4.834, Val R^2=0.416, Val Pearson Coefficient=0.654
Prior best: 0.5866291656576353
Current best: 0.5807160812205282
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=5.05, R²=0.389, Pearson r=0.652
Epoch [26/150]: Train Loss=0.5710, Val Loss=0.6157, Val MAE=5.052, Val R^2=0.389, Val Pearson Coefficient=0.652


Evaluation: MAE=5.86, R²=0.250, Pearson r=0.647
Epoch [27/150]: Train Loss=0.5417, Val Loss=0.7457, Val MAE=5.863, Val R^2=0.250, Val Pearson Coefficient=0.647


Evaluation: MAE=5.26, R²=0.348, Pearson r=0.650
Epoch [28/150]: Train Loss=0.5734, Val Loss=0.6498, Val MAE=5.264, Val R^2=0.348, Val Pearson Coefficient=0.650


Evaluation: MAE=4.90, R²=0.307, Pearson r=0.650
Epoch [29/150]: Train Loss=0.5688, Val Loss=0.7199, Val MAE=4.904, Val R^2=0.307, Val Pearson Coefficient=0.650


Evaluation: MAE=5.81, R²=0.046, Pearson r=0.639
Epoch [30/150]: Train Loss=0.5490, Val Loss=1.0024, Val MAE=5.811, Val R^2=0.046, Val Pearson Coefficient=0.639


Evaluation: MAE=5.96, R²=0.217, Pearson r=0.649
Epoch [31/150]: Train Loss=0.5494, Val Loss=0.7819, Val MAE=5.965, Val R^2=0.217, Val Pearson Coefficient=0.649


Evaluation: MAE=4.83, R²=0.401, Pearson r=0.645
Epoch [32/150]: Train Loss=0.5471, Val Loss=0.6035, Val MAE=4.830, Val R^2=0.401, Val Pearson Coefficient=0.645


Evaluation: MAE=4.93, R²=0.409, Pearson r=0.658
Epoch [33/150]: Train Loss=0.5597, Val Loss=0.5938, Val MAE=4.929, Val R^2=0.409, Val Pearson Coefficient=0.658


Evaluation: MAE=4.73, R²=0.359, Pearson r=0.649
Epoch [34/150]: Train Loss=0.5433, Val Loss=0.6803, Val MAE=4.727, Val R^2=0.359, Val Pearson Coefficient=0.649


Evaluation: MAE=4.91, R²=0.411, Pearson r=0.658
Epoch [35/150]: Train Loss=0.5582, Val Loss=0.5956, Val MAE=4.912, Val R^2=0.411, Val Pearson Coefficient=0.658


Evaluation: MAE=5.31, R²=0.345, Pearson r=0.655
Epoch [36/150]: Train Loss=0.5508, Val Loss=0.6890, Val MAE=5.310, Val R^2=0.345, Val Pearson Coefficient=0.655


Evaluation: MAE=5.30, R²=0.347, Pearson r=0.654
Epoch [37/150]: Train Loss=0.5587, Val Loss=0.6687, Val MAE=5.303, Val R^2=0.347, Val Pearson Coefficient=0.654


Evaluation: MAE=5.80, R²=0.062, Pearson r=0.626
Epoch [38/150]: Train Loss=0.5461, Val Loss=0.9523, Val MAE=5.801, Val R^2=0.062, Val Pearson Coefficient=0.626


Evaluation: MAE=6.04, R²=0.007, Pearson r=0.639
Epoch [39/150]: Train Loss=0.5486, Val Loss=1.0225, Val MAE=6.043, Val R^2=0.007, Val Pearson Coefficient=0.639


Evaluation: MAE=5.11, R²=0.378, Pearson r=0.648
Epoch [40/150]: Train Loss=0.5438, Val Loss=0.6268, Val MAE=5.107, Val R^2=0.378, Val Pearson Coefficient=0.648


Evaluation: MAE=4.58, R²=0.415, Pearson r=0.652
Epoch [41/150]: Train Loss=0.5477, Val Loss=0.5885, Val MAE=4.581, Val R^2=0.415, Val Pearson Coefficient=0.652


Evaluation: MAE=5.65, R²=0.098, Pearson r=0.639
Epoch [42/150]: Train Loss=0.5284, Val Loss=0.9433, Val MAE=5.653, Val R^2=0.098, Val Pearson Coefficient=0.639


Evaluation: MAE=4.91, R²=0.309, Pearson r=0.628
Epoch [43/150]: Train Loss=0.5335, Val Loss=0.6938, Val MAE=4.914, Val R^2=0.309, Val Pearson Coefficient=0.628


Evaluation: MAE=4.69, R²=0.406, Pearson r=0.638
Epoch [44/150]: Train Loss=0.5463, Val Loss=0.6018, Val MAE=4.691, Val R^2=0.406, Val Pearson Coefficient=0.638


Evaluation: MAE=4.61, R²=0.398, Pearson r=0.646
Epoch [45/150]: Train Loss=0.5284, Val Loss=0.6021, Val MAE=4.612, Val R^2=0.398, Val Pearson Coefficient=0.646


Evaluation: MAE=5.11, R²=0.373, Pearson r=0.651
Epoch [46/150]: Train Loss=0.5388, Val Loss=0.6224, Val MAE=5.113, Val R^2=0.373, Val Pearson Coefficient=0.651


Evaluation: MAE=5.18, R²=0.223, Pearson r=0.627
Epoch [47/150]: Train Loss=0.5513, Val Loss=0.7817, Val MAE=5.179, Val R^2=0.223, Val Pearson Coefficient=0.627


Evaluation: MAE=5.30, R²=0.341, Pearson r=0.649
Epoch [48/150]: Train Loss=0.5379, Val Loss=0.6630, Val MAE=5.297, Val R^2=0.341, Val Pearson Coefficient=0.649


Evaluation: MAE=4.65, R²=0.418, Pearson r=0.646
Epoch [49/150]: Train Loss=0.5437, Val Loss=0.5893, Val MAE=4.649, Val R^2=0.418, Val Pearson Coefficient=0.646


Evaluation: MAE=7.32, R²=-0.118, Pearson r=0.643
Epoch [50/150]: Train Loss=0.5195, Val Loss=1.1142, Val MAE=7.319, Val R^2=-0.118, Val Pearson Coefficient=0.643


Evaluation: MAE=4.91, R²=0.305, Pearson r=0.644
Epoch [51/150]: Train Loss=0.5466, Val Loss=0.7416, Val MAE=4.910, Val R^2=0.305, Val Pearson Coefficient=0.644


Evaluation: MAE=4.76, R²=0.416, Pearson r=0.649
Epoch [52/150]: Train Loss=0.5185, Val Loss=0.5869, Val MAE=4.759, Val R^2=0.416, Val Pearson Coefficient=0.649


Evaluation: MAE=4.59, R²=0.408, Pearson r=0.646
Epoch [53/150]: Train Loss=0.5289, Val Loss=0.5942, Val MAE=4.589, Val R^2=0.408, Val Pearson Coefficient=0.646


Evaluation: MAE=4.73, R²=0.416, Pearson r=0.648
Epoch [54/150]: Train Loss=0.5263, Val Loss=0.5809, Val MAE=4.735, Val R^2=0.416, Val Pearson Coefficient=0.648


Evaluation: MAE=4.63, R²=0.387, Pearson r=0.648
Epoch [55/150]: Train Loss=0.5367, Val Loss=0.6149, Val MAE=4.635, Val R^2=0.387, Val Pearson Coefficient=0.648


Evaluation: MAE=4.67, R²=0.423, Pearson r=0.655
Epoch [56/150]: Train Loss=0.5270, Val Loss=0.5728, Val MAE=4.667, Val R^2=0.423, Val Pearson Coefficient=0.655
Prior best: 0.5807160812205282
Current best: 0.5727901993126705
Best model saved to: saved_models/best_bmi_prediction_model.pth


Evaluation: MAE=4.84, R²=0.393, Pearson r=0.643
Epoch [57/150]: Train Loss=0.5300, Val Loss=0.6192, Val MAE=4.836, Val R^2=0.393, Val Pearson Coefficient=0.643


Evaluation: MAE=4.82, R²=0.409, Pearson r=0.647
Epoch [58/150]: Train Loss=0.5231, Val Loss=0.5959, Val MAE=4.820, Val R^2=0.409, Val Pearson Coefficient=0.647


Evaluation: MAE=4.69, R²=0.366, Pearson r=0.642
Epoch [59/150]: Train Loss=0.5164, Val Loss=0.6295, Val MAE=4.692, Val R^2=0.366, Val Pearson Coefficient=0.642


Evaluation: MAE=4.86, R²=0.396, Pearson r=0.641
Epoch [60/150]: Train Loss=0.5185, Val Loss=0.6060, Val MAE=4.861, Val R^2=0.396, Val Pearson Coefficient=0.641


Evaluation: MAE=5.22, R²=0.216, Pearson r=0.643
Epoch [61/150]: Train Loss=0.5350, Val Loss=0.7761, Val MAE=5.222, Val R^2=0.216, Val Pearson Coefficient=0.643


Evaluation: MAE=4.69, R²=0.418, Pearson r=0.662
Epoch [62/150]: Train Loss=0.5117, Val Loss=0.5797, Val MAE=4.695, Val R^2=0.418, Val Pearson Coefficient=0.662


Evaluation: MAE=5.04, R²=0.383, Pearson r=0.657
Epoch [63/150]: Train Loss=0.5520, Val Loss=0.6276, Val MAE=5.042, Val R^2=0.383, Val Pearson Coefficient=0.657


Evaluation: MAE=4.93, R²=0.392, Pearson r=0.645
Epoch [64/150]: Train Loss=0.5177, Val Loss=0.6307, Val MAE=4.929, Val R^2=0.392, Val Pearson Coefficient=0.645


Evaluation: MAE=4.70, R²=0.394, Pearson r=0.638
Epoch [65/150]: Train Loss=0.5163, Val Loss=0.6104, Val MAE=4.703, Val R^2=0.394, Val Pearson Coefficient=0.638


Evaluation: MAE=4.66, R²=0.385, Pearson r=0.645
Epoch [66/150]: Train Loss=0.5277, Val Loss=0.6110, Val MAE=4.661, Val R^2=0.385, Val Pearson Coefficient=0.645


Evaluation: MAE=5.55, R²=0.264, Pearson r=0.650
Epoch [67/150]: Train Loss=0.5182, Val Loss=0.7716, Val MAE=5.547, Val R^2=0.264, Val Pearson Coefficient=0.650


Evaluation: MAE=4.82, R²=0.365, Pearson r=0.633
Epoch [68/150]: Train Loss=0.5146, Val Loss=0.6469, Val MAE=4.817, Val R^2=0.365, Val Pearson Coefficient=0.633


Evaluation: MAE=4.69, R²=0.399, Pearson r=0.642
Epoch [69/150]: Train Loss=0.5105, Val Loss=0.6044, Val MAE=4.689, Val R^2=0.399, Val Pearson Coefficient=0.642


Evaluation: MAE=5.70, R²=0.084, Pearson r=0.616
Epoch [70/150]: Train Loss=0.5286, Val Loss=0.9131, Val MAE=5.698, Val R^2=0.084, Val Pearson Coefficient=0.616


Evaluation: MAE=4.91, R²=0.301, Pearson r=0.625
Epoch [71/150]: Train Loss=0.5188, Val Loss=0.6922, Val MAE=4.908, Val R^2=0.301, Val Pearson Coefficient=0.625


Evaluation: MAE=4.74, R²=0.411, Pearson r=0.642
Epoch [72/150]: Train Loss=0.5175, Val Loss=0.5822, Val MAE=4.735, Val R^2=0.411, Val Pearson Coefficient=0.642


Evaluation: MAE=6.65, R²=0.030, Pearson r=0.640
Epoch [73/150]: Train Loss=0.5094, Val Loss=0.9968, Val MAE=6.650, Val R^2=0.030, Val Pearson Coefficient=0.640


Evaluation: MAE=5.95, R²=0.196, Pearson r=0.636
Epoch [74/150]: Train Loss=0.5208, Val Loss=0.7982, Val MAE=5.949, Val R^2=0.196, Val Pearson Coefficient=0.636


Evaluation: MAE=4.79, R²=0.404, Pearson r=0.638
Epoch [75/150]: Train Loss=0.5213, Val Loss=0.6081, Val MAE=4.795, Val R^2=0.404, Val Pearson Coefficient=0.638


Evaluation: MAE=4.97, R²=0.297, Pearson r=0.617
Epoch [76/150]: Train Loss=0.5051, Val Loss=0.7051, Val MAE=4.966, Val R^2=0.297, Val Pearson Coefficient=0.617


Evaluation: MAE=6.20, R²=0.161, Pearson r=0.625
Epoch [77/150]: Train Loss=0.5262, Val Loss=0.8345, Val MAE=6.205, Val R^2=0.161, Val Pearson Coefficient=0.625


Evaluation: MAE=5.39, R²=0.313, Pearson r=0.644
Epoch [78/150]: Train Loss=0.5126, Val Loss=0.6966, Val MAE=5.388, Val R^2=0.313, Val Pearson Coefficient=0.644


Evaluation: MAE=5.12, R²=0.371, Pearson r=0.640
Epoch [79/150]: Train Loss=0.5110, Val Loss=0.6409, Val MAE=5.118, Val R^2=0.371, Val Pearson Coefficient=0.640


Evaluation: MAE=4.66, R²=0.385, Pearson r=0.641
Epoch [80/150]: Train Loss=0.4948, Val Loss=0.6237, Val MAE=4.662, Val R^2=0.385, Val Pearson Coefficient=0.641


Evaluation: MAE=5.26, R²=0.313, Pearson r=0.642
Epoch [81/150]: Train Loss=0.4972, Val Loss=0.6836, Val MAE=5.256, Val R^2=0.313, Val Pearson Coefficient=0.642


Evaluation: MAE=5.18, R²=0.239, Pearson r=0.618
Epoch [82/150]: Train Loss=0.5138, Val Loss=0.7522, Val MAE=5.176, Val R^2=0.239, Val Pearson Coefficient=0.618


Evaluation: MAE=4.86, R²=0.379, Pearson r=0.628
Epoch [83/150]: Train Loss=0.5090, Val Loss=0.6415, Val MAE=4.861, Val R^2=0.379, Val Pearson Coefficient=0.628


Evaluation: MAE=4.80, R²=0.355, Pearson r=0.623
Epoch [84/150]: Train Loss=0.4985, Val Loss=0.7040, Val MAE=4.800, Val R^2=0.355, Val Pearson Coefficient=0.623


Evaluation: MAE=5.21, R²=0.220, Pearson r=0.625
Epoch [85/150]: Train Loss=0.4922, Val Loss=0.7899, Val MAE=5.215, Val R^2=0.220, Val Pearson Coefficient=0.625


Evaluation: MAE=5.48, R²=0.135, Pearson r=0.605
Epoch [86/150]: Train Loss=0.4988, Val Loss=0.8926, Val MAE=5.478, Val R^2=0.135, Val Pearson Coefficient=0.605


Evaluation: MAE=5.42, R²=0.162, Pearson r=0.623
Epoch [87/150]: Train Loss=0.5085, Val Loss=0.8382, Val MAE=5.416, Val R^2=0.162, Val Pearson Coefficient=0.623


Evaluation: MAE=6.35, R²=0.107, Pearson r=0.631
Epoch [88/150]: Train Loss=0.4965, Val Loss=0.9166, Val MAE=6.350, Val R^2=0.107, Val Pearson Coefficient=0.631


Evaluation: MAE=4.85, R²=0.332, Pearson r=0.635
Epoch [89/150]: Train Loss=0.4905, Val Loss=0.6719, Val MAE=4.845, Val R^2=0.332, Val Pearson Coefficient=0.635


Evaluation: MAE=4.88, R²=0.392, Pearson r=0.635
Epoch [90/150]: Train Loss=0.4974, Val Loss=0.6095, Val MAE=4.879, Val R^2=0.392, Val Pearson Coefficient=0.635


Evaluation: MAE=4.93, R²=0.372, Pearson r=0.640
Epoch [91/150]: Train Loss=0.4970, Val Loss=0.6371, Val MAE=4.931, Val R^2=0.372, Val Pearson Coefficient=0.640


Evaluation: MAE=4.88, R²=0.387, Pearson r=0.628
Epoch [92/150]: Train Loss=0.5054, Val Loss=0.6179, Val MAE=4.883, Val R^2=0.387, Val Pearson Coefficient=0.628


Evaluation: MAE=5.44, R²=0.141, Pearson r=0.617
Epoch [93/150]: Train Loss=0.4960, Val Loss=0.9047, Val MAE=5.445, Val R^2=0.141, Val Pearson Coefficient=0.617


Evaluation: MAE=5.03, R²=0.355, Pearson r=0.634
Epoch [94/150]: Train Loss=0.4933, Val Loss=0.6485, Val MAE=5.033, Val R^2=0.355, Val Pearson Coefficient=0.634


Evaluation: MAE=4.91, R²=0.301, Pearson r=0.627
Epoch [95/150]: Train Loss=0.4922, Val Loss=0.6964, Val MAE=4.910, Val R^2=0.301, Val Pearson Coefficient=0.627


Evaluation: MAE=4.78, R²=0.405, Pearson r=0.639
Epoch [96/150]: Train Loss=0.4977, Val Loss=0.5867, Val MAE=4.782, Val R^2=0.405, Val Pearson Coefficient=0.639


Evaluation: MAE=4.71, R²=0.407, Pearson r=0.644
Epoch [97/150]: Train Loss=0.4771, Val Loss=0.5947, Val MAE=4.709, Val R^2=0.407, Val Pearson Coefficient=0.644


Evaluation: MAE=4.82, R²=0.329, Pearson r=0.622
Epoch [98/150]: Train Loss=0.4946, Val Loss=0.6906, Val MAE=4.818, Val R^2=0.329, Val Pearson Coefficient=0.622


Evaluation: MAE=5.13, R²=0.236, Pearson r=0.625
Epoch [99/150]: Train Loss=0.4914, Val Loss=0.7918, Val MAE=5.127, Val R^2=0.236, Val Pearson Coefficient=0.625


Evaluation: MAE=4.84, R²=0.394, Pearson r=0.635
Epoch [100/150]: Train Loss=0.4855, Val Loss=0.6089, Val MAE=4.838, Val R^2=0.394, Val Pearson Coefficient=0.635


Evaluation: MAE=4.71, R²=0.407, Pearson r=0.639
Epoch [101/150]: Train Loss=0.4817, Val Loss=0.5926, Val MAE=4.707, Val R^2=0.407, Val Pearson Coefficient=0.639


Evaluation: MAE=4.72, R²=0.382, Pearson r=0.621
Epoch [102/150]: Train Loss=0.4793, Val Loss=0.6141, Val MAE=4.719, Val R^2=0.382, Val Pearson Coefficient=0.621


Evaluation: MAE=5.13, R²=0.323, Pearson r=0.639
Epoch [103/150]: Train Loss=0.4672, Val Loss=0.7025, Val MAE=5.133, Val R^2=0.323, Val Pearson Coefficient=0.639


Evaluation: MAE=4.79, R²=0.399, Pearson r=0.635
Epoch [104/150]: Train Loss=0.4809, Val Loss=0.6046, Val MAE=4.788, Val R^2=0.399, Val Pearson Coefficient=0.635


Evaluation: MAE=5.21, R²=0.332, Pearson r=0.630
Epoch [105/150]: Train Loss=0.4793, Val Loss=0.7094, Val MAE=5.210, Val R^2=0.332, Val Pearson Coefficient=0.630


Evaluation: MAE=4.93, R²=0.375, Pearson r=0.617
Epoch [106/150]: Train Loss=0.4926, Val Loss=0.6547, Val MAE=4.926, Val R^2=0.375, Val Pearson Coefficient=0.617


Evaluation: MAE=4.72, R²=0.362, Pearson r=0.633
Epoch [107/150]: Train Loss=0.4745, Val Loss=0.6499, Val MAE=4.717, Val R^2=0.362, Val Pearson Coefficient=0.633


Evaluation: MAE=4.98, R²=0.281, Pearson r=0.635
Epoch [108/150]: Train Loss=0.4955, Val Loss=0.7100, Val MAE=4.982, Val R^2=0.281, Val Pearson Coefficient=0.635


Evaluation: MAE=5.28, R²=0.330, Pearson r=0.642
Epoch [109/150]: Train Loss=0.4859, Val Loss=0.6696, Val MAE=5.275, Val R^2=0.330, Val Pearson Coefficient=0.642


Evaluation: MAE=5.66, R²=0.245, Pearson r=0.609
Epoch [110/150]: Train Loss=0.4718, Val Loss=0.7877, Val MAE=5.665, Val R^2=0.245, Val Pearson Coefficient=0.609


Evaluation: MAE=4.83, R²=0.328, Pearson r=0.619
Epoch [111/150]: Train Loss=0.4847, Val Loss=0.6975, Val MAE=4.835, Val R^2=0.328, Val Pearson Coefficient=0.619


Evaluation: MAE=5.95, R²=0.009, Pearson r=0.596
Epoch [112/150]: Train Loss=0.4911, Val Loss=1.0117, Val MAE=5.951, Val R^2=0.009, Val Pearson Coefficient=0.596


Evaluation: MAE=4.82, R²=0.333, Pearson r=0.617
Epoch [113/150]: Train Loss=0.4752, Val Loss=0.6645, Val MAE=4.821, Val R^2=0.333, Val Pearson Coefficient=0.617


Evaluation: MAE=5.29, R²=0.311, Pearson r=0.624
Epoch [114/150]: Train Loss=0.4726, Val Loss=0.6832, Val MAE=5.294, Val R^2=0.311, Val Pearson Coefficient=0.624


Evaluation: MAE=6.11, R²=0.109, Pearson r=0.626
Epoch [115/150]: Train Loss=0.4793, Val Loss=0.8964, Val MAE=6.108, Val R^2=0.109, Val Pearson Coefficient=0.626


Evaluation: MAE=4.88, R²=0.345, Pearson r=0.628
Epoch [116/150]: Train Loss=0.4713, Val Loss=0.6495, Val MAE=4.884, Val R^2=0.345, Val Pearson Coefficient=0.628


Evaluation: MAE=5.12, R²=0.329, Pearson r=0.624
Epoch [117/150]: Train Loss=0.4773, Val Loss=0.6769, Val MAE=5.117, Val R^2=0.329, Val Pearson Coefficient=0.624


Evaluation: MAE=4.87, R²=0.322, Pearson r=0.622
Epoch [118/150]: Train Loss=0.4830, Val Loss=0.6978, Val MAE=4.873, Val R^2=0.322, Val Pearson Coefficient=0.622


Evaluation: MAE=4.84, R²=0.388, Pearson r=0.623
Epoch [119/150]: Train Loss=0.4652, Val Loss=0.6096, Val MAE=4.835, Val R^2=0.388, Val Pearson Coefficient=0.623


Evaluation: MAE=4.83, R²=0.339, Pearson r=0.607
Epoch [120/150]: Train Loss=0.4606, Val Loss=0.6784, Val MAE=4.829, Val R^2=0.339, Val Pearson Coefficient=0.607


Evaluation: MAE=4.83, R²=0.384, Pearson r=0.620
Epoch [121/150]: Train Loss=0.4864, Val Loss=0.6241, Val MAE=4.832, Val R^2=0.384, Val Pearson Coefficient=0.620


Evaluation: MAE=5.09, R²=0.252, Pearson r=0.628
Epoch [122/150]: Train Loss=0.4752, Val Loss=0.7606, Val MAE=5.093, Val R^2=0.252, Val Pearson Coefficient=0.628


Evaluation: MAE=4.75, R²=0.370, Pearson r=0.616
Epoch [123/150]: Train Loss=0.4768, Val Loss=0.6230, Val MAE=4.750, Val R^2=0.370, Val Pearson Coefficient=0.616


Evaluation: MAE=4.70, R²=0.390, Pearson r=0.630
Epoch [124/150]: Train Loss=0.4816, Val Loss=0.6074, Val MAE=4.697, Val R^2=0.390, Val Pearson Coefficient=0.630


Evaluation: MAE=4.75, R²=0.371, Pearson r=0.628
Epoch [125/150]: Train Loss=0.4655, Val Loss=0.6578, Val MAE=4.749, Val R^2=0.371, Val Pearson Coefficient=0.628


Evaluation: MAE=5.53, R²=0.123, Pearson r=0.608
Epoch [126/150]: Train Loss=0.4752, Val Loss=0.8726, Val MAE=5.529, Val R^2=0.123, Val Pearson Coefficient=0.608


Evaluation: MAE=5.17, R²=0.235, Pearson r=0.607
Epoch [127/150]: Train Loss=0.4650, Val Loss=0.7752, Val MAE=5.169, Val R^2=0.235, Val Pearson Coefficient=0.607


Evaluation: MAE=4.72, R²=0.382, Pearson r=0.626
Epoch [128/150]: Train Loss=0.4750, Val Loss=0.6164, Val MAE=4.724, Val R^2=0.382, Val Pearson Coefficient=0.626


Evaluation: MAE=4.79, R²=0.393, Pearson r=0.628
Epoch [129/150]: Train Loss=0.4728, Val Loss=0.6093, Val MAE=4.794, Val R^2=0.393, Val Pearson Coefficient=0.628


Evaluation: MAE=4.96, R²=0.363, Pearson r=0.621
Epoch [130/150]: Train Loss=0.4758, Val Loss=0.6372, Val MAE=4.960, Val R^2=0.363, Val Pearson Coefficient=0.621


Evaluation: MAE=4.74, R²=0.351, Pearson r=0.631
Epoch [131/150]: Train Loss=0.4616, Val Loss=0.6616, Val MAE=4.739, Val R^2=0.351, Val Pearson Coefficient=0.631


Evaluation: MAE=4.68, R²=0.372, Pearson r=0.630
Epoch [132/150]: Train Loss=0.4736, Val Loss=0.6242, Val MAE=4.683, Val R^2=0.372, Val Pearson Coefficient=0.630


Evaluation: MAE=4.77, R²=0.371, Pearson r=0.628
Epoch [133/150]: Train Loss=0.4689, Val Loss=0.6258, Val MAE=4.767, Val R^2=0.371, Val Pearson Coefficient=0.628


Evaluation: MAE=4.81, R²=0.333, Pearson r=0.616
Epoch [134/150]: Train Loss=0.4701, Val Loss=0.6701, Val MAE=4.811, Val R^2=0.333, Val Pearson Coefficient=0.616


Evaluation: MAE=4.75, R²=0.353, Pearson r=0.621
Epoch [135/150]: Train Loss=0.4801, Val Loss=0.6403, Val MAE=4.755, Val R^2=0.353, Val Pearson Coefficient=0.621


Evaluation: MAE=4.76, R²=0.373, Pearson r=0.635
Epoch [136/150]: Train Loss=0.4527, Val Loss=0.6346, Val MAE=4.763, Val R^2=0.373, Val Pearson Coefficient=0.635


Evaluation: MAE=4.85, R²=0.357, Pearson r=0.617
Epoch [137/150]: Train Loss=0.4591, Val Loss=0.6440, Val MAE=4.851, Val R^2=0.357, Val Pearson Coefficient=0.617


Evaluation: MAE=5.13, R²=0.245, Pearson r=0.612
Epoch [138/150]: Train Loss=0.4681, Val Loss=0.7576, Val MAE=5.126, Val R^2=0.245, Val Pearson Coefficient=0.612


Evaluation: MAE=5.07, R²=0.263, Pearson r=0.609
Epoch [139/150]: Train Loss=0.4559, Val Loss=0.7345, Val MAE=5.069, Val R^2=0.263, Val Pearson Coefficient=0.609


Evaluation: MAE=4.79, R²=0.360, Pearson r=0.613
Epoch [140/150]: Train Loss=0.4529, Val Loss=0.7401, Val MAE=4.788, Val R^2=0.360, Val Pearson Coefficient=0.613


Evaluation: MAE=4.87, R²=0.380, Pearson r=0.623
Epoch [141/150]: Train Loss=0.4628, Val Loss=0.6286, Val MAE=4.866, Val R^2=0.380, Val Pearson Coefficient=0.623


Evaluation: MAE=5.02, R²=0.337, Pearson r=0.635
Epoch [142/150]: Train Loss=0.4548, Val Loss=0.6616, Val MAE=5.019, Val R^2=0.337, Val Pearson Coefficient=0.635


Evaluation: MAE=5.12, R²=0.343, Pearson r=0.600
Epoch [143/150]: Train Loss=0.4412, Val Loss=0.6839, Val MAE=5.121, Val R^2=0.343, Val Pearson Coefficient=0.600


Evaluation: MAE=5.02, R²=0.354, Pearson r=0.623
Epoch [144/150]: Train Loss=0.4553, Val Loss=0.6412, Val MAE=5.018, Val R^2=0.354, Val Pearson Coefficient=0.623


Evaluation: MAE=4.72, R²=0.382, Pearson r=0.628
Epoch [145/150]: Train Loss=0.4471, Val Loss=0.6140, Val MAE=4.720, Val R^2=0.382, Val Pearson Coefficient=0.628


Evaluation: MAE=4.77, R²=0.362, Pearson r=0.626
Epoch [146/150]: Train Loss=0.4440, Val Loss=0.6380, Val MAE=4.773, Val R^2=0.362, Val Pearson Coefficient=0.626


Evaluation: MAE=5.40, R²=0.283, Pearson r=0.610
Epoch [147/150]: Train Loss=0.4569, Val Loss=0.7172, Val MAE=5.398, Val R^2=0.283, Val Pearson Coefficient=0.610


Evaluation: MAE=4.81, R²=0.360, Pearson r=0.612
Epoch [148/150]: Train Loss=0.4395, Val Loss=0.6380, Val MAE=4.810, Val R^2=0.360, Val Pearson Coefficient=0.612


Evaluation: MAE=4.74, R²=0.382, Pearson r=0.622
Epoch [149/150]: Train Loss=0.4587, Val Loss=0.6174, Val MAE=4.736, Val R^2=0.382, Val Pearson Coefficient=0.622


Evaluation: MAE=4.81, R²=0.332, Pearson r=0.620
Epoch [150/150]: Train Loss=0.4601, Val Loss=0.6877, Val MAE=4.811, Val R^2=0.332, Val Pearson Coefficient=0.620
Training Complete


In [74]:
# Results from testing set 
model = BMIMultitaskModel(extra_feature_dim=len(train_dataset.extra_feature_cols)).to(DEVICE)
checkpoint = torch.load(MODEL_PATH, map_location=DEVICE)
model.load_state_dict(checkpoint)
model.eval()
_, test_mae, test_r2, test_pearson = evaluate(model, criterion, test_loader)
print(f"\nTesting Set MAE={test_mae:.3f}, R²={test_r2:.3f}, Pearson Coefficient={test_pearson:.3f}")

Evaluation: MAE=4.94, R²=0.439, Pearson r=0.663

Testing Set MAE=4.940, R²=0.439, Pearson Coefficient=0.663


Without widened structure

Evaluation: MAE=4.98, R²=0.432, Pearson r=0.660

Testing Set MAE=4.983, R²=0.432, Pearson Coefficient=0.660


In [75]:
# import torch
# import numpy as np
# from PIL import Image
# import os

# ===============================================
# Load a single image and predict BMI
# ===============================================
def predict_outputs(image_name, model, extra_feature_df, embedding_dir="data/embeddings"):
    model.eval()
    
    embedding_path = os.path.join(embedding_dir, image_name.replace(".bmp", ".npy"))
    if not os.path.exists(embedding_path):
        raise FileNotFoundError(f"Embedding not found for: {image_name}")

    face_tensor = torch.tensor(np.load(embedding_path), dtype=torch.float32).unsqueeze(0)

    row = extra_feature_df[extra_feature_df["name"] == image_name]
    if row.empty:
        raise ValueError(f"No extra feature data for {image_name}")

    extra_cols = train_dataset.extra_feature_cols
    extra_features = torch.tensor(row[extra_cols].values.astype(np.float32), dtype=torch.float32)

    with torch.no_grad():
        pred_bmi, pred_age, pred_gender_logits = model(face_tensor, extra_features)
        pred_gender = "Male" if torch.sigmoid(pred_gender_logits).item() > 0.5 else "Female"

    pred_bmi = pred_bmi.item() * bmi_std + bmi_mean
    
    return pred_bmi, pred_age.item(), pred_gender

In [76]:
test_image = "img_2.bmp" 

pred_bmi, _, pred_gender = predict_outputs(test_image, model, df_final)
actual_bmi = df_final[df_final['name'] == test_image]['bmi'].iloc[0]
print(f"Predicted BMI: {pred_bmi:.2f}, Gender: {pred_gender}")
print(f"Actual BMI for {test_image}: {actual_bmi:.2f}")

Predicted BMI: 35.48, Gender: Female
Actual BMI for img_2.bmp: 34.97


In [77]:
MODEL_PATH

'saved_models/best_bmi_prediction_model.pth'

In [ ]:
torch.save(model.state_dict(), MODEL_PATH)